lom/src/engine/sequenceHandler.go

```diff
@@ -0,0 +1,832 @@
+}
+
+var seqHandler *SeqHandler_t = nil
+
+func GetSeqHandler() *SeqHandler_t {
+    if seqHandler == nil {
+        LogError("InitSeqHandler is not done yet")
+    }
+    return seqHandler
+}
+
+
+func InitSeqHandler(chTimer chan<- int64) {
+    if chTimer == nil {
+        LogError("Internal error: Nil chan")
+    } else if seqHandler != nil {
+        LogError("Duplicate init seq handler")
+    } else {
+        seqHandler = &SeqHandler_t {
+            activeRequests: make(activeRequestsList_t),
+            sequencesByAnomaly: make(Sequences_t),
+            sequencesByFirstAction: make(Sequences_t),
+            chTimer: chTimer }
+    }
+}
+
+
+/* Called asynchronously from a one shot timer */
+func (p *SeqHandler_t) FireTimer() {
+    if len(p.chTimer) < cap(p.chTimer) {
+        p.chTimer <- 0
+    } 
+    /* Being an alert, as long as there is any outstanding, good enough */
+}
+
+
+/*
+ * Engine calls from main loop upon signal from FireTimer which is
+ * triggered from OneShotTimer fired for seq expiry.
+ * Hence only single routine calls ProcessResponse via
+ * response from action, or processTimeout via chTimer signal
+ * or dropRequest via de-register action process.
+ *
+ * The timeout could be from current active request or overall seq
+ * timeout, whichever expected to fire early.
+ *
+ * In either case, abort sequence.
+ * NOTE: The pending request, if any is *not* dropped from active requests
+ * until response arrives, as Engine is not expected to send a request to
+ * a plugin, while last req is still pending. Hence entry in active requests
+ * only cleared upon action response via client or upon de-register action
+ * by client,
+ */
+func (p *SeqHandler_t) processTimeout() {
+    /*
+     * Complete seq will remove from p.sortedSequencesByDue. Hence always use
+     * current first entry
+     */
+    for len(p.sortedSequencesByDue) > 0 {
+        seq := p.sortedSequencesByDue[0]
+        if time.Now().Unix() >= seq.ExpiryEpoch() {
+            p.completeSequence(seq, LoMSequenceTimeout, "From Process timeout")
+            p.resumeNextSequence()
+        } else {
+            break   /* List is sorted. Hence break */
+        }
+    }
+
+}
+
+
+/*
+ * Called upon action registration by client or upon sequence completion
+ *
+ * Raise request, if this is first action in any configured sequence
+ *
+ * In case of mis-behaving request, PluginMgr will watch & disable the
+ * action, if it returns too frequently.
+ *
+ * Engine need not care.
+ */
+func (p *SeqHandler_t) RaiseRequest(action string) error {
+    regF := GetRegistrations()
+
+    /* Is action registered */
+    if  regF.GetActiveActionInfo(action) == nil {
+        return LogError("Internal: Failing to get active action info (%s)", action)
+    }
+
+    /* Is any active request for this action */
+    if _, ok := p.activeRequests[action]; ok {
+        return LogError("Internal: An active request exists (%s)", action)
+    }
+
+    cfg := GetConfigMgr()
```

Comment:
This part should be before the previous if check. First we will need to check if the action is start of a sequence.

Reply:
We bail out for either, if already active or not first.  The order does not matter.